keras_tuner/tuners/genetic.py

```diff
@@ -0,0 +1,542 @@
+        self.values = {hp.name: hp.default for hp in self.get_space().space}
+        self.parent_1, self.parent_2 = None, None
+
+    @property
+    def _make_max_trials(self):
+        """Calculate the maximum number of trials."""
+        return (
+            self.population_size
+            + self.generation_size * 2 * self.offspring_size
+        )
+
+    @property
+    def _make_ranges(self):
+        """Make the ranges for genetic optimization with respect to max trial.
+
+        Due to the fact that 'oracle' refers to the 'max_trial',
+        the ranges for the genetic optimization have to be calculated
+        for:
+
+            1.  'population_range': the number of models in the
+                population.
+            2.  'offspring_ranges': the number of offspring to produce
+                at each generation.
+            3. 'second_parent_range': the number of models to select
+            4.  'generation_range': the number of generations, where the
+                best models are selected and the offspring are produced.
+
+        Note, 'second_parent_range' is part of the 'offspring_ranges',
+        however, the second evaluation of the 'offspring_ranges' has
+        to be calculated separately, because only one value can be
+        returned once.
+
+        Returns:
+            A dict of four lists: 'population_range', 'generation_range',
+             'offspring_range', 'second_parent_range'.
+        """
+
+        population_range = list(range(self.population_size))
+        generation_range = list(
+            range(
+                self.population_size + self.offspring_size * 2,
+                self.max_trials,
+                self.offspring_size * 2 - 1,
+            )
+        )
+        offspring_range = list(range(self.population_size, self.max_trials))
+        first_parent_range = offspring_range[::2]
+        second_parent_range = offspring_range[1::2]
+        return {
+            "population": population_range,
+            "generation": generation_range,
+            "first_parent": first_parent_range,
+            "second_parent": second_parent_range,
+        }
+
+    def _make_gep(self):
+        """Make a genetic evolutionary process.
+
+        Returns:
+            A `GeneticEvolutionaryProcess` instance.
+        """
+        return GeneticEvolutionaryProcess(
+            mutation_factor=self.mutation_factor,
+            crossover_factor=self.crossover_factor,
+            seed=self.seed,
+        )
+
+    def _check_score(self, score):
+        """Check if the current score is better than the best threshold.
+
+        Args:
+            score: The current score.
+
+        Returns:
+            A `dict` with the status and the current value.
+        """
+        if self.threshold is not None and score <= self.threshold:
+            return {
+                "status": trial_module.TrialStatus.COMPLETED,
+                "values": self.values,
+            }
```

Comment:
The return value is not used?

Reply:
It's true

Reply:
This is actually needed for an early stop ...